fireshape/objective.py

```diff
@@ -1,37 +1,70 @@
-# import ROL
-import _ROL as ROL
+import ROL
 import firedrake as fd
+from .control import ControlSpace, ControlVector
+from .pde_constraint import PdeConstraint
 
-class Objective(ROL.Objective):
 
-    def __init__(self, q, cb=None):
+class Objective(ROL.Objective):
+    "Abstract class of shape functionals."
+    def __init__(self, Q: ControlSpace, cb=None, scale=1.0):
         super().__init__()
-        self.V_m = q.V()
-        self.q = q
-        self.Q = q.controlspace
-        self.cb = cb
+        self.Q = Q # ControlSpace
+        self.V_r = Q.V_r # fd.VectorFunctionSpace on reference mesh
+        self.V_m = Q.V_m # clone of V_r of physical mesh
+        self.cb = cb # method to store current iterate
+        self.scale = scale # scaling factor (multiplies shape functional)
 
-    def val(self):
+        """
+        Preallocate vectors for derivative in FE space and
+        control space so that they are not created every time
+        the derivative is evaluated.
+        """
+        self.deriv_m = fd.Function(self.V_m)
+        self.deriv_r = fd.Function(self.V_r, val=self.deriv_m)
```

Comment:
does this imply that `self.deriv_r' is copy of `self.deriv_m' and that it is updated whenever the latter is? If it is so, it's very cool!